File: backend-v5/app/schemas/kpi.py
"""KPI Pydantic schemas for Palantir-grade dashboard metrics."""
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class ActiveAlertsKPI(BaseModel):
    """Active alerts requiring immediate attention."""

    count: int = Field(description="Total active alerts (CRITICAL + HIGH severity in last 6h)")
    by_category: dict[str, int] = Field(
        default_factory=dict,
        description="Alert count by category (DISASTER, SECURITY, etc.)",
    )
    by_severity: dict[str, int] = Field(
        default_factory=dict,
        description="Alert count by severity (CRITICAL, HIGH)",
    )
    oldest_unresolved_hours: float = Field(
        default=0.0,
        description="Age of oldest unresolved alert in hours",
    )
    requires_attention: bool = Field(
        default=False,
        description="True if any CRITICAL alerts exist",
    )


class EventsKPI(BaseModel):
    """Event counts for the time window."""

    total: int = Field(description="Total unique events (deduplicated)")
    clustered: int = Field(default=0, description="Events that are part of clusters")
    unclustered: int = Field(default=0, description="Standalone events")
    disaster_incidents: int = Field(default=0, description="BIPAD disaster incidents")
    change_vs_yesterday: float = Field(
        default=0.0,
        description="Percentage change vs previous day",
    )
    trend: str = Field(
        default="STABLE",
        description="INCREASING | STABLE | DECREASING",
    )


class ThreatLevelKPI(BaseModel):
    """Composite threat assessment."""

    level: str = Field(description="CRITICAL | ELEVATED | GUARDED | LOW")
    score: float = Field(
        ge=0,
        le=100,
        description="Normalized threat score 0-100",
    )
    trajectory: str = Field(
        default="STABLE",
        description="ESCALATING | STABLE | DE-ESCALATING",
    )
    primary_driver: str = Field(
        default="",
        description="Category causing highest score",
    )
    confidence: float = Field(
        ge=0,
        le=1,
        default=1.0,
        description="Confidence based on data volume (0-1)",
    )


class DistrictsKPI(BaseModel):
    """Geographic spread metrics."""

    affected_count: int = Field(description="Number of districts with activity")
    total_districts: int = Field(default=77, description="Total districts in Nepal")
    affected_percentage: float = Field(description="Percentage of districts affected")
    by_province: dict[str, int] = Field(
        default_factory=dict,
        description="Affected districts by province",
    )
    hotspots: list[str] = Field(
        default_factory=list,
        description="Top 3 districts by event count",
    )


class SourceCoverageKPI(BaseModel):
    """Data pipeline health metrics."""

    active_sources: int = Field(description="Sources with data in last hour")
    total_sources: int = Field(description="Total configured sources")
    coverage_percentage: float = Field(description="Active/Total percentage")
    last_fetch_seconds_ago: int = Field(
        default=0,
        description="Seconds since last successful fetch",
    )
    stale_sources: list[str] = Field(
        default_factory=list,
        description="Sources with no data > 1 hour",
    )


class TrendVelocityKPI(BaseModel):
    """Rate of change metrics."""

    events_this_hour: int = Field(description="Events in current hour")
    events_prev_hour: int = Field(description="Events in previous hour")
    change_percentage: float = Field(description="Hour-over-hour change %")
    direction: str = Field(default="STABLE", description="UP | DOWN | STABLE")
    anomaly_detected: bool = Field(
        default=False,
        description="True if >2 std dev from normal",
    )


class CasualtiesKPI(BaseModel):
    """Casualty metrics from disasters."""

    deaths: int = Field(default=0)
    injured: int = Field(default=0)
    missing: int = Field(default=0)
    affected_families: int = Field(default=0)


class EntityMention(BaseModel):
    """Top entity mention."""

    name: str
    mention_count: int
    category: str = Field(default="UNKNOWN", description="PERSON | ORGANIZATION | LOCATION")


class KPISnapshot(BaseModel):
    """Complete KPI state at a point in time."""

    timestamp: datetime = Field(description="When this snapshot was computed")
    data_freshness_seconds: int = Field(
        description="Age of underlying data in seconds",
    )
    time_window_hours: int = Field(description="Time window for KPI computation")

    # Primary KPIs
    active_alerts: ActiveAlertsKPI
    events_today: EventsKPI
    threat_level: ThreatLevelKPI
    districts_affected: DistrictsKPI
    source_coverage: SourceCoverageKPI
    trend_velocity: TrendVelocityKPI

    # Secondary KPIs
    critical_events: int = Field(default=0, description="CRITICAL severity events only")
    casualties_24h: CasualtiesKPI = Field(default_factory=CasualtiesKPI)
    economic_impact_npr: float = Field(
        default=0.0,
        description="Sum of estimated_loss from disasters",
    )
    top_entities: list[EntityMention] = Field(
        default_factory=list,
        description="Top mentioned entities",
    )


class AlertDetail(BaseModel):
    """Detailed alert for drill-down view."""

    id: UUID
    title: str
    category: str
    severity: str
    source: str
    district: Optional[str] = None
    timestamp: datetime
    url: Optional[str] = None
    summary: Optional[str] = None

    # For disasters
    deaths: Optional[int] = None
    injured: Optional[int] = None
    estimated_loss: Optional[float] = None


class HourlyTrend(BaseModel):
    """Hourly event count for sparkline."""

    hour: datetime
    count: int
    category_breakdown: dict[str, int] = Field(default_factory=dict)

File: frontend/src/types/kpi.ts
// KPI Types for Palantir-grade Dashboard Metrics

export interface ActiveAlertsKPI {
  count: number;
  by_category: Record<string, number>;
  by_severity: Record<string, number>;
  oldest_unresolved_hours: number;
  requires_attention: boolean;
}

export interface EventsKPI {
  total: number;
  clustered: number;
  unclustered: number;
  disaster_incidents: number;
  change_vs_yesterday: number;
  trend: 'INCREASING' | 'STABLE' | 'DECREASING';
}

export interface ThreatLevelKPI {
  level: 'CRITICAL' | 'ELEVATED' | 'GUARDED' | 'LOW';
  score: number;
  trajectory: 'ESCALATING' | 'STABLE' | 'DE-ESCALATING';
  primary_driver: string;
  confidence: number;
}

export interface DistrictsKPI {
  affected_count: number;
  total_districts: number;
  affected_percentage: number;
  by_province: Record<string, number>;
  hotspots: string[];
}

export interface SourceCoverageKPI {
  active_sources: number;
  total_sources: number;
  coverage_percentage: number;
  last_fetch_seconds_ago: number;
  stale_sources: string[];
}

export interface TrendVelocityKPI {
  events_this_hour: number;
  events_prev_hour: number;
  change_percentage: number;
  direction: 'UP' | 'DOWN' | 'STABLE';
  anomaly_detected: boolean;
}

export interface CasualtiesKPI {
  deaths: number;
  injured: number;
  missing: number;
  affected_families: number;
}

export interface EntityMention {
  name: string;
  mention_count: number;
  category: string;
}

export interface KPISnapshot {
  timestamp: string;
  data_freshness_seconds: number;
  time_window_hours: number;

  // Primary KPIs
  active_alerts: ActiveAlertsKPI;
  events_today: EventsKPI;
  threat_level: ThreatLevelKPI;
  districts_affected: DistrictsKPI;
  source_coverage: SourceCoverageKPI;
  trend_velocity: TrendVelocityKPI;

  // Secondary KPIs
  critical_events: number;
  casualties_24h: CasualtiesKPI;
  economic_impact_npr: number;
  top_entities: EntityMention[];
}

export interface AlertDetail {
  id: string;
  title: string;
  category: string;
  severity: string;
  source: string;
  district?: string;
  timestamp: string;
  url?: string;
  summary?: string;
  deaths?: number;
  injured?: number;
  estimated_loss?: number;
}

export interface HourlyTrend {
  hour: string;
  count: number;
  category_breakdown: Record<string, number>;
}
